index.bs

```diff
@@ -111,32 +186,70 @@ dictionary MediaStreamTrackProcessorInit {
+8. Set |processor|.`[[isClosed]]` to false.
+9. Return |processor|.
 
 ### Attributes ### {#attributes-processor}
 <dl>
 <dt><dfn for=MediaStreamTrackProcessor>readable</dfn></dt>
-<dd>Allows reading the frames flowing through the {{MediaStreamTrack}} stored
-in the `[[track]]` internal slot. If `[[track]]`
-is a video track, chunks read from {{MediaStreamTrackProcessor/readable}} will be {{VideoFrame}}
-objects. If `[[track]]` is an audio track, chunks read from
-{{MediaStreamTrackProcessor/readable}} will produce {{AudioFrame}} objects.
-If media frames are not read from {{MediaStreamTrackProcessor/readable}} quickly enough,
-the {{MediaStreamTrackProcessor}} will internally buffer up to `[[maxBufferSize]]`
-of the frames produced by `[[track]]`. If the internal buffer
-is full, each time `[[track]]` produces a new frame, the oldest frame
-in the buffer MUST be dropped and the new frame MUST be added to the buffer.
+<dd>Allows reading the frames delivered by the {{MediaStreamTrack}} stored
+in the `[[track]]` internal slot. This attribute is created the first time it is invoked
+according to the following steps:
+1. Initialize [=this=].{{MediaStreamTrackProcessor/readable}} to be a new {{ReadableStream}}.
+2. <a dfn for="ReadableStream">Set up</a> [=this=].{{MediaStreamTrackProcessor/readable}} with its [=ReadableStream/set up/pullAlgorithm=] set to [=processorPull=] with |this| as parameter, [=ReadableStream/set up/cancelAlgorithm=] set to [=processorCancel=] with |this| as parameter, and [=ReadableStream/set up/highWatermark=] set to 0.
+
+The <dfn>processorPull</dfn> algorithm is given a |processor| as input. It is defined by the following steps:
+1. Increment the value of the |processor|.`[[numPendingReads]]` by 1.
+2. [=Queue a task=] to run the [=maybeReadFrame=] algorithm with |processor| as parameter.
+3. Return  [=a promise resolved with=] undefined.
+
+The <dfn>maybeReadFrame</dfn> algorithm is given a |processor| as input. It is defined by the following steps:
+1. If |processor|.`[[queue]]` is {{Queue/empty}}, abort these steps.
+2. If |processor|.`[[numPendingReads]]` equals zero, abort these steps.
+3. {{Queue/dequeue}} a frame from |processor|.`[[queue]]` and [=ReadableStream/Enqueue=] it in |processor|.{{MediaStreamTrackProcessor/readable}}.
+4. Decrement |processor|.`[[numPendingReads]]` by 1.
+5. Go to step 1.
+
+The <dfn>processorCancel</dfn> algorithm is given a |processor| as input.
+It is defined by running the following steps:
+1. Run the [=processorClose=] algorithm.
+3. Return  [=a promise resolved with=] undefined.
+
+The <dfn>processorClose</dfn> algorithm is given a |processor| as input.
+It is defined by running the following steps:
+1. If |processor|.`[[isClosed]]` is true, abort these steps.
+2. Disconnect |processor| from |processor|.`[[track]]`. The mechanism to do this is UA specific and the result is that |processor| is no longer a sink of |processor|.`[[track]]`.
```

Comment:
Adding or removing a sink does not affect the track's readyState.
For sinks there is no difference if the track is ended or not, AFAICT.
For an ended track, I guess it's the same if it has sinks or not since no media is delivered to the sinks. mediacapture-main does not say anything in this regard.